library/ziti.c

```diff
@@ -1783,3 +1796,93 @@ void ziti_queue_work(ziti_context ztx, ztx_work_f w, void *data) {
 
     uv_async_send(&ztx->w_async);
 }
+
+int ziti_context_init(ziti_context *ztx, const ziti_config *config) {
+    if (config == NULL || config->controller_url == NULL) {
+        ZITI_LOG(ERROR, "config or controller/tls has to be set");
+        return ZITI_INVALID_CONFIG;
+    }
+
+    ziti_context ctx = calloc(1, sizeof(*ctx));
+
+    if (strncmp(config->id.ca, "file://", strlen("file://")) == 0) {
+        struct tlsuv_url_s url;
+        if (tlsuv_parse_url(&url, config->id.ca) != 0) {
+            ZITI_LOG(ERROR, "invalid CA bundle reference[]");
+            return ZITI_INVALID_CONFIG;
+        }
+
+        char *ca = NULL;
+        size_t ca_len;
+        int rc = load_file(url.path, url.path_len, &ca, &ca_len);
+        if (rc == 0) {
+            FREE(ctx->config.id.ca);
+            ctx->config.id.ca = ca;
+        }
+    } else {
+        ctx->config.id.ca = strdup(config->id.ca);
+    }
+
+    ctx->config.controller_url = strdup(config->controller_url);
+    if (config->id.key) ctx->config.id.key = strdup(config->id.key);
+    if (config->id.cert) ctx->config.id.cert = strdup(config->id.cert);
+
+    ctx->opts = default_options;
+
+    *ztx = ctx;
+    return ZITI_OK;
+}
+
+
+int ziti_context_set_options(ziti_context ztx, const ziti_options *options) {
+    if (options == NULL) {
+        ztx->opts = default_options;
+    } else {
+#define copy_opt(f) if (options->f != 0) ztx->opts.f = options->f
```

Comment:
what's the point of this macro? I'd think it would be OK for the 0 values to be assigned as well?

Reply:
the goal is to allow partial updates, without accidentally whacking values set from `default_options`
for example, the app may not care about the `api_page_size` so it won't set it in the struct.